### PR TITLE
fix performance problem for invalid "avatarEntityData"

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* fix a bug where invalid avatarEntityData could kill remote interface performance

### CODE REVIEW

File: interface/src/avatar/Avatar.cpp [hunk 1]
```diff
@@ -295,9 +292,7 @@ void Avatar::updateAvatarEntities() {
         }
     });
 
-    if (success) {
-        setAvatarEntityDataChanged(false);
-    }
+    setAvatarEntityDataChanged(false);
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: These changes in `Avatar::updateAvatarEntities()` are the actual **fix** for the bug.  The problem was that the `avatarEntityData` sent through the AvatarMixer was invalid such that at least one of the entities was never created, which caused these "avatar entities data" to be unpacked and applied every frame --> the interface would run out of time to update the render geometry of other avatars.

File: interface/src/avatar/Avatar.cpp [hunk 2]
```diff
@@ -364,6 +359,9 @@ void Avatar::simulate(float deltaTime, bool inView) {
         measureMotionDerivatives(deltaTime);
         simulateAttachments(deltaTime);
         updatePalms();
+    }
+    {
+        PROFILE_RANGE(simulation, "entities");
         updateAvatarEntities();
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: The offending contex has its own trace data point so we can troubleshoot this easier in the future.

File: interface/src/avatar/MyAvatar.cpp [hunk 3]
```diff
@@ -376,7 +376,9 @@ void MyAvatar::update(float deltaTime) {
         Q_ARG(glm::vec3, (getPosition() - halfBoundingBoxDimensions)),
         Q_ARG(glm::vec3, (halfBoundingBoxDimensions*2.0f)));
 
-    if (_avatarEntityDataLocallyEdited) {
+    uint64_t now = usecTimestampNow();
+    if (now > _identityPacketExpiry || _avatarEntityDataLocallyEdited) {
+        _identityPacketExpiry = now + AVATAR_IDENTITY_PACKET_SEND_INTERVAL_MSECS;
         sendIdentityPacket();
     }
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: Not related to the bug per-se but something I noticed: we would sent the **AvatarIdentity** packet on a timer and then would also send it when things changed which would not reset the timer.  We now pretty much only ever send it in `MyAvatar::update()` and we compute the next expiry when we do.

File: libraries/avatars/src/AvatarData.cpp [hunk 4]
```diff
@@ -2245,17 +2243,15 @@ void AvatarData::setAvatarEntityData(const AvatarEntityMap& avatarEntityData) {
...
-            foreach (auto entityID, previousAvatarEntityIDs) {
-                if (!_avatarEntityData.contains(entityID)) {
-                    _avatarEntityDetached.insert(entityID);
-                }
+        foreach (auto entityID, previousAvatarEntityIDs) {
+            if (!_avatarEntityData.contains(entityID)) {
+                _avatarEntityDetached.insert(entityID);
             }
         }
     });
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.cpp]: By making `AvatarData::setAvatarEntityData()` private we can dispense with most of the sanity checking logic at its beginning.  Also, this method should not be exposed to JS anyway.

File: libraries/entities/src/EntityEditPacketSender.cpp [hunk 5]
```diff
@@ -93,7 +81,8 @@ void EntityEditPacketSender::queueEditEntityMessage(PacketType type,
         return; // bail early
     }
 
-    if (properties.getClientOnly()) {
+    if (properties.getClientOnly() && properties.getOwningAvatarID() == _myAvatar->getID()) {
+        // this is an avatar-based entity --> update our avatar-data rather than sending to the entity-server
         queueEditAvatarEntityMessage(type, entityTree, entityItemID, properties);
         return;
     }
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityEditPacketSender.cpp]: Similarly, by making `EntityEditPacketSender::queueEditAvatarEntityMessage()` private we can dispense with some sanity checking logic.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4276d0a2769b88fc95c7ed6b83e3ac338cfe803d available at: https://gist.github.com/a1b6ab5a11a14357d214c96379d1024b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9642-4276d0a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9642-4276d0a.dmg

**Comment by AndrewMeadows** [comment]
The problem here was: invalid "avatar entity data" was being passed through the AvatarMixer.  At least one of the entities could not be created on recipient interface clients and they would constantly retry every frame, which was CPU expensive.

The context where the CPU cycles were being spent was time-boxed to try to maintain high frame rates when trying to render many avatars --> the time would all get used up for processing the avatar entities and the render geometry of other avatars would never get updated.

To test:

(1) You need three avatars to test: UserA, UserB, UserC.

(2) UserA must hack their Interface.json settings file to send bogus avatarEntityData through the AvatarMixer like so:

(a) Find the local file. On windows it would be something like:
`C:\Users\YOUR_NAME\AppData\Roaming\High Fidelity - dev\Interface.json`
(b) Edit the Interface.json file with a text editor and remove any /Avatar/avatarEntityData lines from it.
(c) In their place paste in the text in the attached file which adds several entity items that are not valid
(d) Save the Interface.json file.

(3) Login all three users with UserA standing between UserB and UserC.

(4) With a bugged interface client UserB and UserC will not be able to see each other's avatars move or animate, but they can see changes to UserA.  UserA can see changes in UserB and UserC.

(5) With this PR all users should be able to see the other two change position and animate.

**Comment by AndrewMeadows** [comment]
It has been pointed out that these changes conflict with the efforts in #9611 which was trying to expose the AvatarEntityData to JS as a valid feature.  Hrm... what to do?

**Comment by AndrewMeadows** [comment]
Don't merge this yet.  After thinking it over I've decided that we should go ahead and expose the setAvatarEntityData() to JS but better protect other interfaces from bad data.  In particular the Interface should:

(1) flag data entries as bad after it fails to produce a viable entity so the Interface doesn't constantly try to parse it.
(2) recognize good data entries that have already been processed in earlier IdentityPackets.
(3) enforce a maximum number of AvatarEntityData entries (maybe a bytesize limit?): this feature cannot remain unbounded.

File: libraries/avatars/src/AvatarData.cpp [hunk 6]
```diff
@@ -2202,14 +2200,24 @@ void AvatarData::setAttachmentsVariant(const QVariantList& variant) {
...
+        if (itr == _avatarEntityData.end()) {
+            if (_avatarEntityData.size() < MAX_NUM_AVATAR_ENTITIES) {
+                _avatarEntityData.insert(entityID, entityData);
+                _avatarEntityDataLocallyEdited = true;
+            }
+        } else {
+            itr.value() = entityData;
+            _avatarEntityDataLocallyEdited = true;
+        }
     });
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.cpp]: We now cap the total number of "avatar entities".

File: interface/src/avatar/Avatar.cpp [hunk 7]
```diff
@@ -228,17 +228,43 @@ void Avatar::updateAvatarEntities() {
...
+                    // we will try again just in case something changed (unlikely).
+                    // Unfortunately constantly trying to build the entity for this data costs
+                    // CPU cycles that we'd rather not spend.
+                    // TODO? put a maximum number of tries on this?
+                }
+            } else {
+                // remember this hash for the future
+                stateItr = _avatarEntityDataHashes.insert(entityID, AvatarEntityDataHash(newHash));
+            }
+            ++dataItr;
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: We now remember each "avatar entity" entry (by its hash) and don't bother parsing entries that have already been handled successfully.  This saves CPU cycles that were being consumed every AvatarIdentityPacket when someone was suing the "avatar entity" feature.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 66a315cb9cfb38efe906c0f014c98793887e863b available at: https://gist.github.com/a1b6ab5a11a14357d214c96379d1024b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9642-66a315c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9642-66a315c.dmg

**Comment by nbq** [comment]
Testing with @Nex-Pro and @claudioshg

**Comment by nbq** [comment]
Tester: Coal
System: i7 6700k, GeForce 970, Windows 10

* Using Piper and Claudioshg, we each tested this build.
* Coal being UserA hacked up the interface file (included below for reference) to included invalid entries from the pastebin provided.
* Saved Interface.json and logged back in and stood between the other two testers.
* We each tried logging in and out and closing our interface and coming back in to see if we triggered any freezes - it did not.
* We were all able to see each others fluid movements and no lock ups so this is a:

**PASS**

**Pull request merged**